libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/pwmout_api.c

```diff
@@ -0,0 +1,226 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "mbed_assert.h"
+#include "pwmout_api.h"
+#include "cmsis.h"
+#include "pinmap.h"
+#include "mbed_error.h"
+
+// Ported from LPC824 and adapted.
+
+#if DEVICE_PWMOUT
+
+#define PWM_IRQn     SCT_IRQn
+
+// Bit flags for used SCT Outputs
+static unsigned char sct_used = 0;
+static int sct_inited = 0;
+
+// Find available output channel
+// Max number of PWM outputs is 4 on LPC812
+static int get_available_sct() {
+   int i;
+       
+// Find available output channel 0..3
+// Also need one Match register per channel
+    for (i = 0; i < CONFIG_SCT_nOU; i++) {
+//    for (i = 0; i < 4; i++) {
+        if ((sct_used & (1 << i)) == 0)
+            return i;
+    }
+    return -1;
+}
+
+// Any Port pin may be used for PWM.
+// Max number of PWM outputs is 4
+void pwmout_init(pwmout_t* obj, PinName pin) {
+    MBED_ASSERT(pin != (uint32_t)NC);
+
+    int sct_n = get_available_sct();
+    if (sct_n == -1) {
+        error("No available SCT Output");
+    }
+
+    sct_used |= (1 << sct_n);
+
+    obj->pwm =  (LPC_SCT_TypeDef*)LPC_SCT;
+    obj->pwm_ch = sct_n;
+
+    LPC_SCT_TypeDef* pwm = obj->pwm;
+
+    // Init SCT on first use
+    if (! sct_inited) {
```

Comment:
Hi Martin,

 

it doesnt look like it is possible to reliably retrieve free outputs/SCT
counters by reading the hardware. The SCT does not support an ''idle''
register as is available for the MRT.

I considered checking the timer reload value and assuming the timer is free
when the value is zero. However, user code may want to set the value to zero
to temporarily disable the PWM. So this also needs admin overhead for
reliable detection. The current solution is similar to what I have seen for
peripherals such as I2C on other mbed platforms. It is also the same
solution that is already used for the current LPC824 mbed PWM library.

 

Wim

  _____  

From: Martin Kojtal [mailto:notifications@github.com] 
Sent: donderdag 12 februari 2015 9:36
To: mbedmicro/mbed
Cc: Wim
Subject: Re: [mbed] Create pwmout_api.c (#906)

 

In libraries/mbed/targets/hal/TARGET_NXP/TARGET_LPC81X/pwmout_api.c
<https://github.com/mbedmicro/mbed/pull/906#discussion_r24565305> :

> +    MBED_ASSERT(pin != (uint32_t)NC);
> +
> +    int sct_n = get_available_sct();
> +    if (sct_n == -1) {
> +        error("No available SCT Output");
> +    }
> +
> +    sct_used |= (1 << sct_n);
> +
> +    obj->pwm =  (LPC_SCT_TypeDef*)LPC_SCT;
> +    obj->pwm_ch = sct_n;
> +
> +    LPC_SCT_TypeDef* pwm = obj->pwm;
> +
> +    // Init SCT on first use
> +    if (! sct_inited) {

Can be this retrieved from hw, as a run time check, no variable involved?

-
Reply to this email directly or view it on
<https://github.com/mbedmicro/mbed/pull/906/files#r24565305>  GitHub.
<https://github.com/notifications/beacon/AKMUNCTSFyXMdkv1XzRHMPwBcIwhXNG1ks5
nrF2BgaJpZM4DfF7f.gif>